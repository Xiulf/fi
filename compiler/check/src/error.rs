use crate::subst::Subst;
use crate::ty::Ty;
use hir::ir::Span;

pub type UnifyResult = Result<Subst, UnifyError>;

pub enum UnifyError {
    Mismatch,
}

pub enum TypeError {
    Mismatched {
        a_ty: Ty,
        a_span: Span,
        b_ty: Ty,
        b_span: Span,
    },
}

impl TypeError {
    pub fn report(self, file: source::FileId, db: &dyn crate::TypeDatabase) {
        use diagnostics::Label;

        match self {
            TypeError::Mismatched {
                a_ty,
                a_span,
                b_ty,
                b_span,
            } => db
                .to_diag_db()
                .error(format!(
                    "mismatched types: `{}` != `{}`",
                    a_ty.display(db),
                    b_ty.display(db)
                ))
                .with_label(Label::primary(file, a_span))
                .with_label(Label::secondary(file, b_span))
                .finish(),
        }
    }
}
