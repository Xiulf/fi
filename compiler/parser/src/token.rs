use std::fmt;

use text_size::TextSize;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, serde::Serialize, serde::Deserialize)]
pub struct Token {
    pub kind: SyntaxKind,
    pub len: TextSize,
}

#[allow(non_camel_case_types)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, serde::Serialize, serde::Deserialize)]
#[repr(u16)]
pub enum SyntaxKind {
    ERROR,
    // EOF,
    WHITESPACE,
    COMMENT,

    LYT_START,
    LYT_SEP,
    LYT_END,

    IDENT,
    TYPE,
    SYMBOL,
    INT,
    FLOAT,
    CHAR,
    STRING,

    L_PAREN,
    R_PAREN,
    L_BRACE,
    R_BRACE,
    L_BRACKET,
    R_BRACKET,

    DOT,
    COMMA,
    EQUALS,
    UNDERSCORE,
    ARROW,
    LEFT_ARROW,
    DBL_COLON,
    DBL_DOT,
    PIPE,
    AT,
    TICK,

    MODULE_KW,
    IMPORT_KW,
    HIDING_KW,
    INFIXL_KW,
    INFIXR_KW,
    INFIX_KW,
    PREFIX_KW,
    POSTFIX_KW,
    FOREIGN_KW,
    CONST_KW,
    STATIC_KW,
    TYPE_KW,
    TRAIT_KW,
    IMPL_KW,
    AS_KW,
    DO_KW,
    TRY_KW,
    FN_KW,
    IF_KW,
    THEN_KW,
    ELSE_KW,
    MATCH_KW,
    WITH_KW,
    WHERE_KW,
    LET_KW,
    RECUR_KW,
    RETURN_KW,

    SOURCE_FILE,

    ATTR,
    ATTR_ARGS,
    ATTR_ARG_IDENT,
    ATTR_ARG_CALL,
    ATTR_ARG_EQUAL,
    ATTR_ARG_LIT,

    ITEM_MODULE,
    ITEM_IMPORT,
    ITEM_FIXITY,
    ITEM_VALUE,
    ITEM_TYPE,
    ITEM_TRAIT,
    ITEM_IMPL,

    EXPORTS,
    EXPORT_NAME,
    EXPORT_MODULE,

    IMPORT_ITEMS,
    IMPORT_ITEM,
    IMPORT_HIDING,

    TYPE_VARS,

    CTOR,
    CTOR_RECORD,
    CTOR_FIELD,

    WHERE_CLAUSE,
    WHERE_CLAUSE_CONSTRAINT,
    WHERE_CLAUSE_VAR_KIND,

    TYPE_PARENS,
    TYPE_WHERE,
    TYPE_FUNC,
    TYPE_LIST,
    TYPE_APP,
    TYPE_INFIX,
    TYPE_PATH,
    TYPE_VAR,
    TYPE_HOLE,
    TYPE_UNIT,

    PAT_TYPED,
    PAT_PARENS,
    PAT_UNIT,
    PAT_WILDCARD,
    PAT_LITERAL,
    PAT_BIND,
    PAT_PATH,
    PAT_APP,
    PAT_INFIX,

    EXPR_TYPED,
    EXPR_PARENS,
    EXPR_HOLE,
    EXPR_UNIT,
    EXPR_PATH,
    EXPR_LITERAL,
    EXPR_LAMBDA,
    EXPR_INFIX,
    EXPR_PREFIX,
    EXPR_POSTFIX,
    EXPR_APP,
    EXPR_METHOD,
    EXPR_BLOCK,
    EXPR_TRY,
    EXPR_IF,
    EXPR_MATCH,
    EXPR_RETURN,

    MATCH_ARM,
    MATCH_GUARD,

    STMT_LET,
    STMT_BIND,
    STMT_EXPR,

    LIT_INT,
    LIT_FLOAT,
    LIT_CHAR,
    LIT_STRING,

    PATH,
    PATH_SEGMENT,

    NAME,
    NAME_REF,

    __LAST,
}

impl SyntaxKind {
    pub fn is_trivia(self) -> bool {
        matches!(self, SyntaxKind::WHITESPACE | SyntaxKind::COMMENT)
    }
}

impl fmt::Display for SyntaxKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            | Self::ERROR => "{error}",
            | Self::WHITESPACE => "whitespace",
            | Self::COMMENT => "comment",
            | Self::LYT_START => "indent",
            | Self::LYT_SEP => "newline",
            | Self::LYT_END => "unindent",
            | Self::IDENT => "identifier",
            | Self::TYPE => "type name",
            | Self::SYMBOL => "symbol",
            | Self::INT => "int literal",
            | Self::FLOAT => "float literal",
            | Self::CHAR => "character literal",
            | Self::STRING => "string literal",
            | Self::L_PAREN => "'('",
            | Self::R_PAREN => "')'",
            | Self::L_BRACE => "'{'",
            | Self::R_BRACE => "'}'",
            | Self::L_BRACKET => "'['",
            | Self::R_BRACKET => "']'",
            | Self::DOT => "'.'",
            | Self::COMMA => "','",
            | Self::EQUALS => "'='",
            | Self::UNDERSCORE => "'_'",
            | Self::ARROW => "'->'",
            | Self::LEFT_ARROW => "'<-'",
            | Self::DBL_COLON => "'::'",
            | Self::DBL_DOT => "'..'",
            | Self::PIPE => "'|'",
            | Self::AT => "'@'",
            | Self::TICK => "'`'",
            | Self::MODULE_KW => "'module'",
            | Self::IMPORT_KW => "'import'",
            | Self::HIDING_KW => "'hiding'",
            | Self::INFIXL_KW => "'infixl'",
            | Self::INFIXR_KW => "'infixr'",
            | Self::INFIX_KW => "'infix'",
            | Self::PREFIX_KW => "'prefix'",
            | Self::POSTFIX_KW => "'postfix'",
            | Self::FOREIGN_KW => "'foreign'",
            | Self::CONST_KW => "'const'",
            | Self::STATIC_KW => "'static'",
            | Self::TYPE_KW => "'type'",
            | Self::TRAIT_KW => "'trait'",
            | Self::IMPL_KW => "'impl'",
            | Self::AS_KW => "'as'",
            | Self::DO_KW => "'do'",
            | Self::TRY_KW => "'try'",
            | Self::FN_KW => "'fn'",
            | Self::IF_KW => "'if'",
            | Self::THEN_KW => "'then'",
            | Self::ELSE_KW => "'else'",
            | Self::MATCH_KW => "'match'",
            | Self::WITH_KW => "'with'",
            | Self::WHERE_KW => "'where'",
            | Self::LET_KW => "'let'",
            | Self::RECUR_KW => "'recur'",
            | Self::RETURN_KW => "'return'",
            | Self::__LAST => unreachable!(),
            | _ => return write!(f, "{self:?}"),
        };

        f.write_str(s)
    }
}
