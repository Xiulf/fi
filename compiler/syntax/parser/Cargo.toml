[package]
name = "parser"
version = "0.2.0"
authors = ["Cyberduc-k <cyberduck_@outlook.com>"]
edition = "2018"

[dependencies]
codespan = "0.9.5"
codespan-reporting = "0.9.5"
unicode-xid = "0.2.0"
