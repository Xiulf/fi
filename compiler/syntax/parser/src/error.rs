use codespan::FileId;
use codespan_reporting::diagnostic::Diagnostic;

pub type Result<T> = std::result::Result<T, Diagnostic<FileId>>;
