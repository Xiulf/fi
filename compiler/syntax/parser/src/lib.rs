pub mod attr;
pub mod buffer;
pub mod error;
pub mod ident;
pub mod lexer;
pub mod literal;
pub mod parse;
pub mod punct;
pub mod token;

pub fn parse<T: parse::Parse<D>, D>(
    source: &str,
    file: diagnostics::FileId,
    reporter: &diagnostics::Reporter,
    start: Option<diagnostics::Span>,
    data: D,
) -> T {
    use diagnostics::Spanned;

    let mut lexer = lexer::Lexer::new(source, file, reporter);
    let buffer = lexer.run();
    let stream = parse::ParseBuffer::new(
        buffer.begin(),
        reporter,
        data,
        if let Some(start) = start {
            start
        } else if !buffer.tokens.is_empty() {
            buffer.tokens[0].span()
        } else {
            Default::default()
        },
    );

    if let Ok(res) = stream.parse() {
        if !stream.is_empty() {}

        res
    } else {
        unreachable!();
    }
}

// pub fn parse_buffer<T: parse::Parse>(buffer: &buffer::TokenBuffer, flags: usize, start: Option<diagnostics::Span>) -> error::Result<T> {
//     let mut stream = parse::ParseBuffer::new(buffer.begin(), if let Some(start) = start {
//         start
//     } else if !buffer.tokens.is_empty() {
//         buffer.tokens[0].span()
//     } else {
//         Default::default()
//     });

//     stream.flags = flags;

//     stream.parse()
// }
