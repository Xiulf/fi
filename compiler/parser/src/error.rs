use std::collections::HashSet;
use std::fmt;
use std::ops::Range;

use chumsky::Error;
use text_size::{TextRange, TextSize};

use crate::token::SyntaxKind;

#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize)]
pub enum SyntaxError {
    ParseError(ParseError),
}

impl fmt::Display for SyntaxError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            | Self::ParseError(e) => e.fmt(f),
        }
    }
}

/// A type representing possible reasons for an error.
#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize)]
pub enum ErrorReason {
    /// An unexpected input was found.
    Unexpected,
    /// An unclosed delimiter was found.
    Unclosed {
        /// The span of the unclosed delimiter.
        span: TextRange,
        /// The unclosed delimiter.
        delimiter: SyntaxKind,
    },
    /// An error with a custom message occurred.
    Custom(String),
}

impl fmt::Display for ErrorReason {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        const DEFAULT_DISPLAY_UNEXPECTED: &str = "unexpected input";

        match self {
            | Self::Unexpected => write!(f, "{}", DEFAULT_DISPLAY_UNEXPECTED),
            | Self::Unclosed { span, delimiter } => {
                write!(
                    f,
                    "unclosed delimiter ({}..{}) in {}",
                    usize::from(span.start()),
                    usize::from(span.end()),
                    delimiter
                )
            },
            | Self::Custom(string) => write!(f, "error {}", string),
        }
    }
}

/// A type representing zero, one, or many labels applied to an error
#[derive(Clone, Copy, Debug, PartialEq, serde::Serialize)]
enum ErrorLabel {
    Some(&'static str),
    None,
    Multi,
}

impl ErrorLabel {
    fn merge(self, other: Self) -> Self {
        match (self, other) {
            | (ErrorLabel::Some(a), ErrorLabel::Some(b)) if a == b => ErrorLabel::Some(a),
            | (ErrorLabel::Some(_), ErrorLabel::Some(_)) => ErrorLabel::Multi,
            | (ErrorLabel::Multi, _) => ErrorLabel::Multi,
            | (_, ErrorLabel::Multi) => ErrorLabel::Multi,
            | (ErrorLabel::None, x) => x,
            | (x, ErrorLabel::None) => x,
        }
    }
}

impl From<ErrorLabel> for Option<&'static str> {
    fn from(label: ErrorLabel) -> Self {
        match label {
            | ErrorLabel::Some(s) => Some(s),
            | _ => None,
        }
    }
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct ParseError {
    span: TextRange,
    reason: ErrorReason,
    expected: HashSet<Option<SyntaxKind>>,
    found: Option<SyntaxKind>,
    label: ErrorLabel,
}

impl ParseError {
    /// Create an error with a custom error message.
    pub fn custom<M: ToString>(span: Range<usize>, msg: M) -> Self {
        Self {
            span: TextRange::new(TextSize::from(span.start as u32), TextSize::from(span.end as u32)),
            reason: ErrorReason::Custom(msg.to_string()),
            expected: HashSet::default(),
            found: None,
            label: ErrorLabel::None,
        }
    }

    /// Returns the span that the error occurred at.
    pub fn span(&self) -> TextRange {
        self.span.clone()
    }

    /// Returns an iterator over possible expected patterns.
    pub fn expected(&self) -> impl ExactSizeIterator<Item = &Option<SyntaxKind>> + '_ {
        self.expected.iter()
    }

    /// Returns the input, if any, that was found instead of an expected pattern.
    pub fn found(&self) -> Option<&SyntaxKind> {
        self.found.as_ref()
    }

    /// Returns the reason for the error.
    pub fn reason(&self) -> &ErrorReason {
        &self.reason
    }

    /// Returns the error's label, if any.
    pub fn label(&self) -> Option<&'static str> {
        self.label.into()
    }
}

impl Error<SyntaxKind> for ParseError {
    type Span = Range<usize>;
    type Label = &'static str;

    fn expected_input_found<Iter: IntoIterator<Item = Option<SyntaxKind>>>(
        span: Self::Span,
        expected: Iter,
        found: Option<SyntaxKind>,
    ) -> Self {
        Self {
            span: TextRange::new(TextSize::from(span.start as u32), TextSize::from(span.end as u32)),
            reason: ErrorReason::Unexpected,
            expected: expected.into_iter().collect(),
            found,
            label: ErrorLabel::None,
        }
    }

    fn unclosed_delimiter(
        unclosed_span: Self::Span,
        delimiter: SyntaxKind,
        span: Self::Span,
        expected: SyntaxKind,
        found: Option<SyntaxKind>,
    ) -> Self {
        Self {
            span: TextRange::new(TextSize::from(span.start as u32), TextSize::from(span.end as u32)),
            reason: ErrorReason::Unclosed {
                span: TextRange::new(
                    TextSize::from(unclosed_span.start as u32),
                    TextSize::from(unclosed_span.end as u32),
                ),
                delimiter,
            },
            expected: core::iter::once(Some(expected)).collect(),
            found,
            label: ErrorLabel::None,
        }
    }

    fn with_label(mut self, label: Self::Label) -> Self {
        match self.label {
            | ErrorLabel::Some(_) => {},
            | _ => {
                self.label = ErrorLabel::Some(label);
            },
        }
        self
    }

    fn merge(mut self, other: Self) -> Self {
        // TODO: Assert that `self.span == other.span` here?
        self.reason = match (&self.reason, &other.reason) {
            | (ErrorReason::Unclosed { .. }, _) => self.reason,
            | (_, ErrorReason::Unclosed { .. }) => other.reason,
            | _ => self.reason,
        };
        self.label = self.label.merge(other.label);
        for expected in other.expected {
            self.expected.insert(expected);
        }
        self
    }
}

impl PartialEq for ParseError {
    fn eq(&self, other: &Self) -> bool {
        self.span == other.span && self.found == other.found && self.reason == other.reason && self.label == other.label
    }
}

impl Eq for ParseError {
}

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        // TODO: Take `self.reason` into account

        if let Some(found) = &self.found {
            write!(f, "found {:?}", found.to_string())?;
        } else {
            write!(f, "found end of input")?;
        };

        match self.expected.len() {
            | 0 => {}, //write!(f, " but end of input was expected")?,
            | 1 => write!(f, " but expected {}", match self.expected.iter().next().unwrap() {
                | Some(x) => format!("{:?}", x.to_string()),
                | None => "end of input".to_string(),
            },)?,
            | _ => {
                write!(
                    f,
                    " but expected one of {}",
                    self.expected
                        .iter()
                        .map(|expected| match expected {
                            | Some(x) => format!("{:?}", x.to_string()),
                            | None => "end of input".to_string(),
                        })
                        .collect::<Vec<_>>()
                        .join(", ")
                )?;
            },
        }

        Ok(())
    }
}

impl std::error::Error for ParseError {
}
