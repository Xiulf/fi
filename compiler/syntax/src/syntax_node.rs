use cstree::interning::Interner;
use cstree::{GreenNode, GreenNodeBuilder, Language};
use parser::error::SyntaxError;
use parser::token::SyntaxKind;

use crate::Parsed;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Lang {}

impl Language for Lang {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: cstree::SyntaxKind) -> Self::Kind {
        assert!(raw.0 < SyntaxKind::__LAST as u16);
        unsafe { std::mem::transmute(raw.0) }
    }

    fn kind_to_raw(kind: Self::Kind) -> cstree::SyntaxKind {
        cstree::SyntaxKind(kind as u16)
    }

    fn static_text(kind: Self::Kind) -> Option<&'static str> {
        match kind {
            | _ => None,
        }
    }
}

pub type SyntaxNode = cstree::SyntaxNode<Lang>;
pub type SyntaxToken = cstree::SyntaxToken<Lang>;
pub type SyntaxElement = cstree::SyntaxElement<Lang>;

pub struct SyntaxTreeBuilder<'i, I> {
    inner: GreenNodeBuilder<'i, 'i, Lang, I>,
    errors: Vec<SyntaxError>,
}

impl<'i, I: Interner> SyntaxTreeBuilder<'i, I> {
    pub fn new(interner: &'i mut I) -> Self {
        Self {
            inner: GreenNodeBuilder::with_interner(interner),
            errors: Vec::new(),
        }
    }

    pub(crate) fn finish_raw(self) -> (GreenNode, Vec<SyntaxError>) {
        let (node, _) = self.inner.finish();

        (node, self.errors)
    }

    pub fn finish(self) -> Parsed<SyntaxNode> {
        let (green, errors) = self.finish_raw();

        Parsed::new(green, errors)
    }

    pub fn token(&mut self, kind: SyntaxKind, text: &str) {
        self.inner.token(kind, text);
    }

    pub fn start_node(&mut self, kind: SyntaxKind) {
        self.inner.start_node(kind);
    }

    pub fn finish_node(&mut self) {
        self.inner.finish_node();
    }

    pub fn error(&mut self, error: SyntaxError) {
        self.errors.push(error);
    }
}
