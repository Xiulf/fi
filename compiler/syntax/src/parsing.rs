use cstree::interning::Interner;
use cstree::TextSize;
use parser::error::SyntaxError;
use parser::token::SyntaxKind;
use parser::TreeSink;

use crate::syntax_node::{SyntaxNode, SyntaxTreeBuilder};
use crate::Parsed;

pub fn parse_text<I: Interner>(text: &str, interner: &mut I) -> Parsed<SyntaxNode> {
    let (output, errors) = parser::parse(text);
    let mut tree_sink = TextTreeSink::new(text, interner);

    if let Some(event) = output {
        parser::event::process(&mut tree_sink, event);
    } else {
        tree_sink.start_node(SyntaxKind::SOURCE_FILE);
        tree_sink.finish_node();
    }

    for error in errors {
        tree_sink.error(SyntaxError::ParseError(error));
    }

    tree_sink.inner.finish()
}

pub struct TextTreeSink<'t, 'i, I> {
    text: &'t str,
    text_pos: TextSize,
    inner: SyntaxTreeBuilder<'i, I>,
}

impl<'t, 'i, I: Interner> TextTreeSink<'t, 'i, I> {
    pub fn new(text: &'t str, interner: &'i mut I) -> Self {
        Self {
            text,
            text_pos: TextSize::from(0),
            inner: SyntaxTreeBuilder::new(interner),
        }
    }
}

impl<'t, I: Interner> TreeSink for TextTreeSink<'t, '_, I> {
    fn token(&mut self, kind: SyntaxKind, len: TextSize) {
        let text = &self.text[self.text_pos.into()..(self.text_pos + len).into()];
        self.text_pos += len;
        self.inner.token(kind, text);
    }

    fn start_node(&mut self, kind: SyntaxKind) {
        self.inner.start_node(kind);
    }

    fn finish_node(&mut self) {
        self.inner.finish_node();
    }

    fn error(&mut self, error: SyntaxError) {
        self.inner.error(error);
    }
}

#[test]
fn test_parsing() {
    let input = r#"
        module Core.Cmp =

        main = 0

        type X =
            | Y
            | Z

        trait Iterator self it =
            next :: self -> Option it

        impl Iterator Iter Item =
            next self = _
    "#;
    let input = unindent::unindent(input.trim());
    let mut interner = cstree::interning::new_interner();
    let parsed = parse_text(&input, &mut interner);

    insta::assert_debug_snapshot!(parsed);
}
