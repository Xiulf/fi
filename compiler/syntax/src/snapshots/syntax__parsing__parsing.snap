---
source: compiler/syntax/src/parsing.rs
expression: parsed
---
Parsed {
    green: HeaderSlice {
        header: HeaderWithLength {
            header: GreenNodeHead {
                kind: SyntaxKind(
                    56,
                ),
                text_len: 154,
                child_hash: 662970387,
            },
            length: 2,
        },
        slice: [
            HeaderSlice {
                header: HeaderWithLength {
                    header: GreenNodeHead {
                        kind: SyntaxKind(
                            57,
                        ),
                        text_len: 154,
                        child_hash: 1981746356,
                    },
                    length: 51,
                },
                slice: [
                    GreenToken {
                        kind: SyntaxKind(
                            32,
                        ),
                        text: Some(
                            Spur(1),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            8,
                        ),
                        text: Some(
                            Spur(3),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            21,
                        ),
                        text: Some(
                            Spur(4),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            8,
                        ),
                        text: Some(
                            Spur(5),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            23,
                        ),
                        text: Some(
                            Spur(6),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            5,
                        ),
                        text: Some(
                            Spur(7),
                        ),
                    },
                    HeaderSlice {
                        header: HeaderWithLength {
                            header: GreenNodeHead {
                                kind: SyntaxKind(
                                    58,
                                ),
                                text_len: 8,
                                child_hash: 123356734,
                            },
                            length: 5,
                        },
                        slice: [
                            GreenToken {
                                kind: SyntaxKind(
                                    7,
                                ),
                                text: Some(
                                    Spur(8),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    2,
                                ),
                                text: Some(
                                    Spur(2),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    23,
                                ),
                                text: Some(
                                    Spur(6),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    2,
                                ),
                                text: Some(
                                    Spur(2),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    11,
                                ),
                                text: Some(
                                    Spur(9),
                                ),
                            },
                        ],
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            5,
                        ),
                        text: Some(
                            Spur(7),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            40,
                        ),
                        text: Some(
                            Spur(10),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            8,
                        ),
                        text: Some(
                            Spur(11),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            23,
                        ),
                        text: Some(
                            Spur(6),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            4,
                        ),
                        text: Some(
                            Spur(12),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            29,
                        ),
                        text: Some(
                            Spur(13),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            8,
                        ),
                        text: Some(
                            Spur(14),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            5,
                        ),
                        text: Some(
                            Spur(12),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            29,
                        ),
                        text: Some(
                            Spur(13),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            8,
                        ),
                        text: Some(
                            Spur(15),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            6,
                        ),
                        text: Some(
                            Spur(7),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            5,
                        ),
                        text: Some(
                            Spur(16),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            41,
                        ),
                        text: Some(
                            Spur(17),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            8,
                        ),
                        text: Some(
                            Spur(18),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            7,
                        ),
                        text: Some(
                            Spur(19),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            7,
                        ),
                        text: Some(
                            Spur(20),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            23,
                        ),
                        text: Some(
                            Spur(6),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            4,
                        ),
                        text: Some(
                            Spur(12),
                        ),
                    },
                    HeaderSlice {
                        header: HeaderWithLength {
                            header: GreenNodeHead {
                                kind: SyntaxKind(
                                    58,
                                ),
                                text_len: 25,
                                child_hash: 16153092,
                            },
                            length: 11,
                        },
                        slice: [
                            GreenToken {
                                kind: SyntaxKind(
                                    7,
                                ),
                                text: Some(
                                    Spur(21),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    2,
                                ),
                                text: Some(
                                    Spur(2),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    27,
                                ),
                                text: Some(
                                    Spur(22),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    2,
                                ),
                                text: Some(
                                    Spur(2),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    7,
                                ),
                                text: Some(
                                    Spur(19),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    2,
                                ),
                                text: Some(
                                    Spur(2),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    25,
                                ),
                                text: Some(
                                    Spur(23),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    2,
                                ),
                                text: Some(
                                    Spur(2),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    8,
                                ),
                                text: Some(
                                    Spur(24),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    2,
                                ),
                                text: Some(
                                    Spur(2),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    7,
                                ),
                                text: Some(
                                    Spur(20),
                                ),
                            },
                        ],
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            6,
                        ),
                        text: Some(
                            Spur(7),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            5,
                        ),
                        text: Some(
                            Spur(16),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            42,
                        ),
                        text: Some(
                            Spur(25),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            8,
                        ),
                        text: Some(
                            Spur(18),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            8,
                        ),
                        text: Some(
                            Spur(26),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            8,
                        ),
                        text: Some(
                            Spur(27),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            2,
                        ),
                        text: Some(
                            Spur(2),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            23,
                        ),
                        text: Some(
                            Spur(6),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            4,
                        ),
                        text: Some(
                            Spur(12),
                        ),
                    },
                    HeaderSlice {
                        header: HeaderWithLength {
                            header: GreenNodeHead {
                                kind: SyntaxKind(
                                    58,
                                ),
                                text_len: 13,
                                child_hash: 3715232857,
                            },
                            length: 7,
                        },
                        slice: [
                            GreenToken {
                                kind: SyntaxKind(
                                    7,
                                ),
                                text: Some(
                                    Spur(21),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    2,
                                ),
                                text: Some(
                                    Spur(2),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    7,
                                ),
                                text: Some(
                                    Spur(19),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    2,
                                ),
                                text: Some(
                                    Spur(2),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    23,
                                ),
                                text: Some(
                                    Spur(6),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    2,
                                ),
                                text: Some(
                                    Spur(2),
                                ),
                            },
                            GreenToken {
                                kind: SyntaxKind(
                                    24,
                                ),
                                text: Some(
                                    Spur(28),
                                ),
                            },
                        ],
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            6,
                        ),
                        text: Some(
                            Spur(16),
                        ),
                    },
                    GreenToken {
                        kind: SyntaxKind(
                            5,
                        ),
                        text: Some(
                            Spur(16),
                        ),
                    },
                ],
            },
            GreenToken {
                kind: SyntaxKind(
                    1,
                ),
                text: Some(
                    Spur(16),
                ),
            },
        ],
    },
    errors: [],
    _marker: PhantomData<fn() -> cstree::syntax::node::SyntaxNode<syntax::syntax_node::Lang>>,
}
